File: Algorithms/Hash Table/438. Find All Anagrams in a String.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
		vector<int> res;
		if(s.empty() || p.empty()) return res;
		vector<int> hash(256, 0);
		int left = 0, right = 0, count = p.size();
		for(char& c : p)
		{
			hash[c] += 1;
		}
		while(right < s.size())
		{
			if(hash[s[right++]]-- > 0) count -= 1;
			if(count == 0) res.emplace_back(left);	
			if(right - left == p.size() && hash[s[left++]]++ >= 0) count += 1;
		}
		return res;
    }
};

File: Algorithms/Two Pointer/438. Find All Anagrams in a String.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int[] cnt = new int[26];
        List<Integer> res = new ArrayList();
        for(int i = 0; i < p.length(); ++i) {
            cnt[p.charAt(i) - 'a'] += 1;
        }
        int count = p.length();
        int left = 0, right = 0;
        while(right < s.length()) {
            if(cnt[s.charAt(right++) - 'a']-- > 0) --count;
            if(right - left == p.length()) {
                if(count == 0) res.add(left);
                if(cnt[s.charAt(left++) - 'a']++ >= 0) ++count;
            }
        }
        return res;
    }
}
